### PR TITLE
feat: Add userError field to task result

### PR DESCRIPTION
**Pull request by jachym-tousek-keboola**
Jira: https://keboola.atlassian.net/browse/PSGO-679

**Changes:**
- Service errors details are now propagated to task result.

-----------

### CODE REVIEW

**Review by Matovidlo — APPROVED** [review]
LGTM

**Review by hosekpeter — APPROVED** [review]

File: internal/pkg/service/common/task/node_test.go [hunk 1]
```diff
@@ -502,7 +516,7 @@ task/123/my-receiver/my-export/some.task/%s
 {"level":"info","message":"started task","component":"task","task":"123/my-receiver/my-export/some.task/%s","node":"node2"}
 {"level":"debug","message":"lock acquired \"runtime/lock/task/my-lock\"","component":"task","task":"123/my-receiver/my-export/some.task/%s","node":"node2"}
 {"level":"info","message":"some message from the task (2)","component":"task","task":"123/my-receiver/my-export/some.task/%s","node":"node2"}
-{"level":"warn","message":"task failed (%s): some error (2) - unexpected [%s]","component":"task","task":"123/my-receiver/my-export/some.task/%s","node":"node2"}
+{"level":"warn","message":"task failed (%s): no space right on device","component":"task","task":"123/my-receiver/my-export/some.task/%s","node":"node2"}
 {"level":"debug","message":"lock released \"runtime/lock/task/my-lock\"","component":"task","task":"123/my-receiver/my-export/some.task/%s","node":"node2"}
```
Comment (michaljurecko) [inline internal/pkg/service/common/task/node_test.go:520]: This is some funny typo, isn't it? `no space left on device`

File: internal/pkg/service/common/task/node.go [hunk 2]
```diff
@@ -319,6 +320,23 @@ func (n *Node) runTask(logger log.Logger, task Task, cfg Config) (result Result,
...
+		if errWithUserMessage, ok := result.Error.(svcerrors.WithUserMessage); ok {
+			task.UserError.Message = errWithUserMessage.ErrorUserMessage()
+		} else {
+			task.UserError.Message = "Unknown error"
+		}
+
+		if errWithExceptionID, ok := result.Error.(svcerrors.WithExceptionID); ok {
+			task.UserError.ExceptionID = errWithExceptionID.ErrorExceptionID()
+		}
 	}
```
Comment (michaljurecko) [inline internal/pkg/service/common/task/node.go:340]: The exceptionID should be always present in the message + optionally in the `expcetionId` field.

Workflow:
- User gets (in the UI) `something wrong `message with exception ID.
- He contacts our support.
- We see in the logs the full error message, with the same exception ID.

File: internal/pkg/service/common/task/node_test.go (hunk 1 above)
Comment (Matovidlo) [reply internal/pkg/service/common/task/node_test.go:520]: I would resolve it, it is just a test, can contain anything

File: internal/pkg/service/common/task/node.go (hunk 2 above)
Comment (michaljurecko) [reply internal/pkg/service/common/task/node.go:340]: Please see the `httpserver` package as reference, if there is no exceptionID you should generate it, for the log purposes mentioned above. The key is to somehow implement the `workflow`.

File: internal/pkg/service/common/task/node_test.go (hunk 1 above)
Comment (michaljurecko) [reply internal/pkg/service/common/task/node_test.go:520]: :+1: Ok, it was not clear to me whether or not it is some type of error that can be used in production as well.

**Comment by michaljurecko** [comment]
I probably won't find time for this anymore, try to come to an agreement in the team.

File: internal/pkg/service/stream/components.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ func StartComponents(ctx context.Context, serviceScp dependencies.ServiceScope,
 	// Common task scope
 	var taskScp commonDeps.TaskScope
 	if componentsMap[ComponentAPI] {
-		taskScp, err = commonDeps.NewTaskScope(ctx, cfg.NodeID, serviceScp)
+		taskScp, err = commonDeps.NewTaskScope(ctx, cfg.NodeID, api.ExceptionIDPrefix, serviceScp)
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/stream/components.go:41]: @jachym-tousek-keboola Instad of `keboola-stream-api-`, I suggest to use `keboola-stream-`, the constant can be in `components.go`, if you don't have a better place.

------------

The prefix doesn't matter, we could use `keboola-stream-` everywhere and not distinguish between API and background work if you want.

File: internal/pkg/service/common/task/node_test.go [hunk 4]
```diff
@@ -423,6 +424,11 @@ task/123/my-receiver/my-export/some.task/%s
   "node": "node1",
   "lock": "runtime/lock/task/my-lock",
   "error": "some error (1) - expected",
+  "userError": {
+    "name": "unknownError",
+    "message": "Unknown error",
+    "exceptionId": "test-node-%s"
```
Comment (michaljurecko) [inline internal/pkg/service/common/task/node_test.go:430]: `test-node` -> `test-service`, it's a bit confusing

**Pull request merged**